### PR TITLE
fix(): Unable to close last window of the application

### PR DESCRIPTION
**Pull request by omerson**
This PR will close https://github.com/bleenco/bterm/issues/29

### CODE REVIEW

File: src/electron.ts [hunk 1]
```diff
@@ -53,26 +55,20 @@ app.on('ready', () => {
...
-      current.close();
-      current = null;
-    }
+    handleWindowsOnClose();
   });
 });
 
 app.on('browser-window-created', (e: Event, win: Electron.BrowserWindow) => {
   current = win;
+  windows = new Array<Electron.BrowserWindow>()
```
Comment (jkuri) [inline src/electron.ts]: ?

File: src/electron.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@ import { platform } from 'os';
 import { keyboardShortcuts } from './keyboard-shortcuts';
 
 let current: Electron.BrowserWindow = null;
+let windows: Electron.BrowserWindow[] = null;
```
Comment (jkuri) [inline src/electron.ts]: `let windows: Electron.BrowserWindow[] = [];`. and you have array prepared.

File: src/electron.ts [hunk 3]
```diff
@@ -90,6 +89,16 @@ app.on('activate', () => {
   }
 });
 
+function handleWindowsOnClose() {
+  unregisterShortcuts();
+  mainWindowState.saveState(current);
+  let index = windows.findIndex((w) => w.id === current.id);
+  windows.splice(index, 1);
+  current.close();
+  windows.filter(w => w);
```
Comment (jkuri) [inline src/electron.ts]: this does nothing here.

File: src/electron.ts [hunk 4]
```diff
@@ -53,26 +55,20 @@ app.on('ready', () => {
...
-      current = null;
-    }
+    handleWindowsOnClose();
   });
 });
 
 app.on('browser-window-created', (e: Event, win: Electron.BrowserWindow) => {
   current = win;
+  windows = new Array<Electron.BrowserWindow>()
+  if (windows.filter(w => w.id === win.id).length === 0) {
```
Comment (jkuri) [inline src/electron.ts]: `if (!windows.filter(w => w.id === win.id)) { ... }`

File: src/electron.ts (hunk 1 above)
Comment (omerson) [inline src/electron.ts]: The idea is to initiate array since it is declared as null, but I will apply comment below. Thanks.

Comment (jkuri) [inline src/electron.ts]: then you do `windows = [];`.

File: src/electron.ts [hunk 5]
```diff
@@ -90,6 +89,16 @@ app.on('activate', () => {
   }
 });
 
+function handleWindowsOnClose() {
+  unregisterShortcuts();
+  mainWindowState.saveState(current);
+  let index = windows.findIndex((w) => w.id === current.id);
```
Comment (omerson) [inline src/electron.ts]: I will check this as well as last comment.

File: src/electron.ts (hunk 3 above)
Comment (omerson) [inline src/electron.ts]: What exactly? The idea is to reindex the array with `windows.filter(w => w);`

Comment (jkuri) [inline src/electron.ts]: what do you mean with reindex?

Comment (omerson) [inline src/electron.ts]: Make its indexes to go from 0 to length-1.

Comment (jkuri) [inline src/electron.ts]: indexes always go from 0 to length - 1, you never need to "reindex". also, `.filter()` returns an array, I believe you meant `windows = windows.filter(w => w);`, but this would make array empty so not sure what did you mean. even if would need to "reindex" array you woud do it with `windows = windows.map(w => w)`... but again, this is not needed.

Comment (omerson) [inline src/electron.ts]: yes it was my obvious confusion that slice would do indexing like 0, 2, 3 etc. and I was missing filter assignment bas to array `windows = windows.filter(w => w);` Clear now.

File: src/electron.ts (hunk 4 above)
Comment (omerson) [inline src/electron.ts]: This will always be true. 
Thus I am going to use `if (!windows.filter(w => w.id === win.id)[0]) { ... }` ?

Comment (jkuri) [inline src/electron.ts]: sorry, I meant `if (!windows.filter(w => w.id === win.id).length)`.

**Comment by omerson** [comment]
This PR contains one test from the list on #19 

> should open and close multi window of the app

File: e2e/app.e2e.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 const Application = require('spectron').Application;
 const { resolve } = require('path');
+const clone = require('clone');
```
Comment (jkuri) [inline e2e/app.e2e.ts]: do we need this extra dependency? we are trying to limit number of dependencies in package.json to minimum, but if you find out that we need this then also install `@types/clone` and import this as `import * as clone from 'clone';`. that's the typescript way.

File: e2e/app.e2e.ts [hunk 7]
```diff
@@ -151,5 +152,13 @@ describe('bterm launch', function() {
       .then(() => this.app.client.browserWindow.isMaximized())
       .then(result => expect(result).to.be.true);
   });
+
+  it('should open and close multi window of the app', () => {
+    let tmpApp = clone(this.app);
+    return this.app.client.waitUntilWindowLoaded()
+      .then(() => tmpApp.start())
+      .then(() => tmpApp.client.click('.close'))
+      .then(() => wait(1000))
```
Comment (jkuri) [inline e2e/app.e2e.ts]: I think you need to tests here if the app is closed / is null? I cannot see any test here.

File: src/electron.ts [hunk 8]
```diff
@@ -90,6 +81,21 @@ app.on('activate', () => {
...
+    windows.push(win);
+  }
+}
+
+function handleWindowsOnClose() {
+  unregisterShortcuts();
+  mainWindowState.saveState(current);
+  windows = windows.filter(w => w.id !== current.id);
+  current.close();
+  current = windows.pop() || null;
```
Comment (jkuri) [inline src/electron.ts]: `windows.pop()` will remove last item in the array, I believe you want 
```ts
current = windows[window.length - 1] || null;
```

File: e2e/app.e2e.ts [hunk 9]
```diff
@@ -2,6 +2,7 @@ const Application = require('spectron').Application;
 const { resolve } = require('path');
 import * as os from 'os';
 import { expect } from 'chai';
+import * as clone from 'clone';
```
Comment (jkuri) [inline e2e/app.e2e.ts]: uneeded?

Comment (omerson) [inline e2e/app.e2e.ts]: No, it is necessary for clone method.

Comment (jkuri) [inline e2e/app.e2e.ts]: why do we need clone method? just for testing?

File: e2e/app.e2e.ts [hunk 10]
```diff
@@ -151,5 +152,25 @@ describe('bterm launch', function() {
       .then(() => this.app.client.browserWindow.isMaximized())
       .then(result => expect(result).to.be.true);
   });
+
+  it('should open and close multi window of the app', () => {
+    let appToClose = clone(this.app);
```
Comment (jkuri) [inline e2e/app.e2e.ts]: this is not correct though, cloning the object does not create new window... you didn't try running this locally?

File: e2e/app.e2e.ts (hunk 9 above)
Comment (omerson) [inline e2e/app.e2e.ts]: Yes it is just for testing.

File: e2e/app.e2e.ts (hunk 10 above)
Comment (omerson) [inline e2e/app.e2e.ts]: New cloned object creates when started creates new window. Yes, I did try this locally.

File: e2e/app.e2e.ts (hunk 9 above)
Comment (jkuri) [inline e2e/app.e2e.ts]: we don't need any extra dependencies imho. if you want to create multiple instances then export function inside `beforeEach` and call this function to create new window. this is overkill. 
hope you didn't struggling with that since Wednesday, because I cannot see any progress then ~5 falsy tests.
